README.md
# next-js-rpc

- Quickly create type-safe RPC endpoints for Next.JS
- Request validation with Zod
- Client side fetching with React Query

## Install

```bash
npm i zod react-query @jonbilous/next-js-rpc
```

## Next.JS API

```js
import { createHandler } from "@jonbilous/next-js-rpc/";
import { getUserSession } from "utils/ctx";
import db from "utils/db";
import zod from "zod";

const ctx = {
  user: getUserSession,
};

const schema = zod.null();

const getLocations = createHandler({
  url: "/api/functions",
  fn: async (params, ctx) => {
    return db.location.findMany();
  },
  schema,
  ctx,
});

export type LocationQuery = typeof getLocations;

export default getLocations;
```

## Next.JS Page

```js
import { client } from "@jonbilous/next-js-rpc";
import type { GetServerSideProps, NextPage } from "next";
import type { LocationQuery } from "pages/api/functions";
import getLocations from "pages/api/functions";

export const getServerSideProps: GetServerSideProps = async (ctx) => {
  const locations = await getLocations.ssr(null, ctx);
  // import handler and call .ssr() to use on the server
  return { props: { locations } };
};

const Home: NextPage = (props) => {
  const query = client.useQuery < LocationQuery > ("/api/functions", null);
  // import type on the client and pass to useQuery - url, request type and response types will be inferred
  return <div></div>;
};

export default Home;
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See
[Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [1.21.0](https://github.com/jonbilous/next-js-rpc/compare/v1.20.0...v1.21.0) (2022-05-04)


### Features

* **error handling:** throw error after sending status to client ([258431e](https://github.com/jonbilous/next-js-rpc/commit/258431e4adbccfd44dc4c6669125d53914c3e3f9))

# [1.20.0](https://github.com/jonbilous/next-js-rpc/compare/v1.19.0...v1.20.0) (2022-04-30)


### Features

* **client:** identify api errors on client ([0d19d9c](https://github.com/jonbilous/next-js-rpc/commit/0d19d9c40352caaf96d454698500d913402ad205))

# [1.19.0](https://github.com/jonbilous/next-js-rpc/compare/v1.18.1...v1.19.0) (2022-04-27)


### Features

* **ssrprovider/consumer:** remove ([2aa1ae0](https://github.com/jonbilous/next-js-rpc/commit/2aa1ae0810c84ed4bd0e1e45bd66b8bb2d925080))

## [1.18.1](https://github.com/jonbilous/next-js-rpc/compare/v1.18.0...v1.18.1) (2022-04-27)


### Bug Fixes

* **ssrprovider:** serialize all props ([ec404aa](https://github.com/jonbilous/next-js-rpc/commit/ec404aa10c9ab80a793693a812b2eb093c9bbfa1))

# [1.18.0](https://github.com/jonbilous/next-js-rpc/compare/v1.17.4...v1.18.0) (2022-04-27)


### Features

* **ssrprovider:** create ssr provider to serialize data ([b9c33b0](https://github.com/jonbilous/next-js-rpc/commit/b9c33b0849a2c6938c962d83cea1d7d3aba70421))

## [1.17.4](https://github.com/jonbilous/next-js-rpc/compare/v1.17.3...v1.17.4) (2022-04-27)


### Bug Fixes

* **ssrconsumer:** copy props before mutating ([b63e240](https://github.com/jonbilous/next-js-rpc/commit/b63e2403e46ed81be2befb448388699af521510c))

## [1.17.3](https://github.com/jonbilous/next-js-rpc/compare/v1.17.2...v1.17.3) (2022-04-27)


### Bug Fixes

* **ssrconsumer:** fix serialization logic ([e7af069](https://github.com/jonbilous/next-js-rpc/commit/e7af069a53a7aa3c2c5211dda3e9bbab29f610a9))

## [1.17.2](https://github.com/jonbilous/next-js-rpc/compare/v1.17.1...v1.17.2) (2022-04-27)


### Bug Fixes

* **server side fn:** superjson serialize on ssr fn ([b3fe9a3](https://github.com/jonbilous/next-js-rpc/commit/b3fe9a35c0a93d2153004732392cc654503d4ede))

## [1.17.1](https://github.com/jonbilous/next-js-rpc/compare/v1.17.0...v1.17.1) (2022-04-27)


### Bug Fixes

* **peer dependencies:** update peer deps ([8569d71](https://github.com/jonbilous/next-js-rpc/commit/8569d71f82739595f0c9be390b8288a7e5541899))

# [1.17.0](https://github.com/jonbilous/next-js-rpc/compare/v1.16.1...v1.17.0) (2022-04-27)


### Features

* **ssrconsumer:** create ssrconsumer HOC for parsing json ([18b62e1](https://github.com/jonbilous/next-js-rpc/commit/18b62e1eca7391b144ee34c1f34f9700d9d83010))

## [1.16.1](https://github.com/jonbilous/next-js-rpc/compare/v1.16.0...v1.16.1) (2022-04-27)


### Bug Fixes

* **cache:** catch errors when getting item from cache ([94e05ef](https://github.com/jonbilous/next-js-rpc/commit/94e05ef602cd521ec60c35f6097a7ec79d52256e))

# [1.16.0](https://github.com/jonbilous/next-js-rpc/compare/v1.15.1...v1.16.0) (2022-04-27)


### Features

* **error handling:** added HTTPError class ([a765fbc](https://github.com/jonbilous/next-js-rpc/commit/a765fbc39f0ebaf0df3c0630ec87b3763094933f))

## [1.15.1](https://github.com/jonbilous/next-js-rpc/compare/v1.15.0...v1.15.1) (2022-04-24)


### Bug Fixes

* **type imports:** fix type imports ([ea6a990](https://github.com/jonbilous/next-js-rpc/commit/ea6a990c50705c9eba6fad4e3c421abe859d8deb))

# [1.15.0](https://github.com/jonbilous/next-js-rpc/compare/v1.14.0...v1.15.0) (2022-04-24)


### Features

* **structure:** support nested imports ([d828fa3](https://github.com/jonbilous/next-js-rpc/commit/d828fa329148abb914b55336726d55a323350dba))

# [1.14.0](https://github.com/jonbilous/next-js-rpc/compare/v1.13.0...v1.14.0) (2022-04-24)


### Features

* **cache:** ssr function to use cache ([ddbcb09](https://github.com/jonbilous/next-js-rpc/commit/ddbcb092a2f13a2f5823bc21931804b2c6865f65))

# [1.13.0](https://github.com/jonbilous/next-js-rpc/compare/v1.12.0...v1.13.0) (2022-04-24)


### Features

* **caching:** add caching feature ([a8bab6a](https://github.com/jonbilous/next-js-rpc/commit/a8bab6a013b90903a245f84a7f70f391b1b8e0f5))

# [1.12.0](https://github.com/jonbilous/next-js-rpc/compare/v1.11.0...v1.12.0) (2022-04-22)


### Features

* **readme:** updated readme ([ca3a90c](https://github.com/jonbilous/next-js-rpc/commit/ca3a90cd43dc94679f1a4addabbf87b440c91dde))

# [1.11.0](https://github.com/jonbilous/next-js-rpc/compare/v1.10.0...v1.11.0) (2022-04-22)


### Features

* **readme:** update the readme ([cce8454](https://github.com/jonbilous/next-js-rpc/commit/cce8454c5aa8aa4769758071a9fb3760b2f0a253))

# [1.10.0](https://github.com/jonbilous/next-js-rpc/compare/v1.9.0...v1.10.0) (2022-04-22)


### Features

* **server:** add server function as property of handler ([f00e03a](https://github.com/jonbilous/next-js-rpc/commit/f00e03aae11515384925c95c98086d30ed5cad65))

# [1.9.0](https://github.com/jonbilous/next-js-rpc/compare/v1.8.0...v1.9.0) (2022-04-22)


### Features

* **readme:** update readme ([c2ab574](https://github.com/jonbilous/next-js-rpc/commit/c2ab5747ef63ef97bb56d6cbb4b3e331d93e2d9d))

# [1.8.0](https://github.com/jonbilous/next-js-rpc/compare/v1.7.0...v1.8.0) (2022-04-22)


### Features

* **server & client:** no more server code ending up on the clienrt ([6b73ba9](https://github.com/jonbilous/next-js-rpc/commit/6b73ba9c2092800a267bbaad2693a437da5409aa))

# [1.7.0](https://github.com/jonbilous/next-js-rpc/compare/v1.6.0...v1.7.0) (2022-04-22)


### Features

* **exports:** export only server ([9c0cba5](https://github.com/jonbilous/next-js-rpc/commit/9c0cba556aac77bc4806a8e4e4a934be6848e8ba))

# [1.6.0](https://github.com/jonbilous/next-js-rpc/compare/v1.5.0...v1.6.0) (2022-04-22)


### Features


src/index.test.ts
import { expect, test } from "vitest";
import { hello } from ".";

test("returns hello world", () => {
  const result = hello();

  expect(result).toBe("Hello world");
  expect(result).toBeTruthy();
});


src/server/index.test.ts
import { createMocks } from "node-mocks-http";
import { describe, expect, test, vi } from "vitest";
import { createApi, HTTPError } from ".";
import zod, { ZodDate } from "zod";
import { CacheProvider } from "types";
import superjson from "superjson";

const cacheProvider: CacheProvider = {
  defaultTtl: 5000,
  get: async <T extends unknown>(key: string) => {
    return null as T;
  },
  write: async (key) => {},
  flush: async (key) => {},
};

describe("test createHandler", () => {
  test("context works", async () => {
    const testValue = "user";

    const mockCtx = vi.fn(async () => {
      return new Promise<string>((resolve, reject) =>
        setTimeout(() => resolve(testValue), 2000)
      );
    });

    const { createHandler } = createApi({
      cacheProvider,
    });

    const { req, res } = createMocks({ method: "POST" });

    const handler = createHandler({
      url: "/hello",
      fn: async (data, ctx) => {
        return ctx.user;
      },
      schema: zod.any(),
      ctx: {
        user: mockCtx,
      },
    });

    await handler(req, res).catch((err) => {});

    expect(mockCtx).toBeCalledTimes(1);

    const result = superjson.parse(res._getData());

    expect(result).toBe(testValue);
  });

  test("httperror works in handler", async () => {
    const code = 404;
    const error = "it does not work!";

    const { createHandler } = createApi({
      cacheProvider,
    });

    const { req, res } = createMocks({ method: "POST" });

    const handler = createHandler({
      url: "/hello",
      fn: async (data, ctx) => {
        throw new HTTPError(error, code);
      },

      schema: zod.any(),
    });

    await handler(req, res).catch((err) => {});

    const result = superjson.parse<{ error: string }>(res._getData());

    expect(res.statusCode).toBe(code);
    expect(result).toHaveProperty("error");
    expect(result.error).toBe(error);
  });

  test("httperror works in ctx", async () => {
    const code = 404;
    const error = "it does not work!";

    const { createHandler } = createApi({
      cacheProvider,
    });

    const { req, res } = createMocks({ method: "POST" });

    const handler = createHandler({
      url: "/hello",
      fn: async (data, ctx) => {
        return { hello: "world" };
      },
      ctx: {
        user: async () => {
          throw new HTTPError(error, code);
        },
      },
      schema: zod.any(),
    });

    await handler(req, res).catch((err) => {});

    const result = superjson.parse<{ error: string }>(res._getData());

    expect(res.statusCode).toBe(code);
    expect(result).toHaveProperty("error");
    expect(result.error).toBe(error);
  });

  test("error returns 501", async () => {
    const { createHandler } = createApi({ cacheProvider });

    const { req, res } = createMocks({ method: "POST" });

    const handler = createHandler({
      url: "/hello",
      fn: async (data, ctx) => {
        return { hello: "world" };
      },
      ctx: {
        user: async () => {
          throw new Error("!!!!!");
        },
      },
      schema: zod.any(),
    });

    await handler(req, res).catch((err) => {});

    const result = superjson.parse<{ error: string }>(res._getData());

    expect(res.statusCode).toBe(501);

    expect(result).toHaveProperty("error");

    expect(result.error).toBe("Internal server error");
  });

  test("schema validation", async () => {
    const { createHandler } = createApi({ cacheProvider });

    const { req, res } = createMocks({
      method: "POST",
      body: { mello: "world" },
    });

    const handler = createHandler({
      url: "/hello",
      fn: async (data, ctx) => {
        return data;
      },
      schema: zod.object({ hello: zod.string() }),
    });

    await handler(req, res).catch((err) => {});

    const result = superjson.parse<{ error: string }>(res._getData());

    expect(res.statusCode).toBe(404);

    expect(result).toHaveProperty("error");

    expect(result.error).toBe("Input validation error");
  });
});


src/server/index.ts
import { NextApiResponse } from "next";
import type {
  ApiRequest,
  CacheProvider,
  ContextResult,
  HandlerContext,
  HandlerDefinition,
} from "../types";
import zod from "zod";
import superjson from "superjson";

export class HTTPError extends Error {
  constructor(public message: string, public status: number) {
    super(message);
  }
}

export const createApi = ({
  cacheProvider,
}: {
  cacheProvider: CacheProvider;
}) => {
  const createHandler = <
    RequestBody,
    ResponseType,
    Url extends string,
    Ctx extends {
      [key: string]: (ctx: HandlerContext) => any;
    }
  >({
    fn,
    schema,
    ctx,
    cache,
    url,
  }: HandlerDefinition<RequestBody, ResponseType, Ctx, Url>) => {
    const validate = (data: RequestBody) => {
      try {
        return schema ? schema.parse(data) : zod.any().parse(data);
      } catch (err) {
        throw new HTTPError("Input validation error", 404);
      }
    };

    const buildContext = async ({ req, res }: HandlerContext) => {
      const contextResult = {} as Record<keyof Ctx, any>;

      await Promise.all(
        Object.entries(ctx || {}).map(async ([key, fn]) => {
          const result = await fn({ req, res });

          contextResult[key as keyof Ctx] = result;
        })
      );

      return { req, res, ...contextResult };
    };

    const getResult = async (
      data: RequestBody,
      ctx: HandlerContext,
      fn: (
        data: RequestBody,
        ctx: HandlerContext<ContextResult<Ctx>>
      ) => Promise<ResponseType>
    ) => {
      const validatedData = validate(data);

      const handlerCtx = await buildContext({ req: ctx.req, res: ctx.res });

      if (cache) {
        const key = [url, cache.getKey(validatedData, handlerCtx)].join("-");

        const cachedResult = await cacheProvider
          .get<ResponseType>(key)
          .catch((err) => null);

        if (cachedResult) {
          return cachedResult;
        } else {
          const result = await fn(validatedData, handlerCtx);

          cacheProvider.write(
            key,
            result,
            cache.ttl ?? cacheProvider.defaultTtl
          );

          return result;
        }
      }

      return fn(validatedData, handlerCtx);
    };

    const handler = async (
      req: ApiRequest<RequestBody, Url>,
      res: NextApiResponse<ResponseType>
    ) => {
      if (req.method !== "POST") {
        return res
          .status(400)
          .json(superjson.serialize({ error: "Must use POST" }) as any);
      }

      return getResult(req.body, { req, res }, fn)
        .then((result) => {
          return res.status(200).json(superjson.serialize(result) as any);
        })
        .catch((err) => {
          const message =
            err instanceof HTTPError ? err.message : "Internal server error";

          const code = err instanceof HTTPError ? err.status : 501;

          res.status(code).json(superjson.serialize({ error: message }) as any);

          throw err;
        });
    };

    handler.ssr = (
      data: RequestBody,
      { req, res }: HandlerContext
    ): Promise<ResponseType> => {
      return getResult(data, { req, res }, fn);
    };

    return handler;
  };

  return { createHandler };
};
